### PR TITLE
Experimental: small batch boost

### PR DESCRIPTION
**Pull request by EamonNerbonne**
Use plain insert statements for small batches.

Baseline (24c7dbfe1147141d8fe0f21ce6b25295b9268be6): 

<pre><code>
BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19041.508 (2004/?/20H1)
Intel Core i7-8700K CPU 3.70GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=5.0.100-rc.1.20452.10
  [Host] : .NET Core 3.1.8 (CoreCLR 4.700.20.41105, CoreFX 4.700.20.41903), X64 RyuJIT
</code></pre>
<pre><code>Job=InProcess  MaxRelativeError=0.002  OutlierMode=DontRemove  
Toolchain=InProcessNoEmitToolchain  IterationCount=15  LaunchCount=1  
MaxIterationCount=5000  WarmupCount=3  
</code></pre>

<table>
<thead><tr><th>                          Method</th><th>  Mean</th><th>Error</th><th>StdDev</th><th>Gen 0</th><th>Gen 1</th><th>Gen 2</th><th>Allocated</th>
</tr>
</thead><tbody><tr><td>BulkInsertTarget_LoadFromTable</td><td>89.511 μs</td><td>0.8805 μs</td><td>0.8236 μs</td><td>1.0986</td><td>-</td><td>-</td><td>7.46 KB</td>
</tr><tr><td>BulkCopyToSqlServer_cachedTarget_0</td><td>2.681 μs</td><td>0.0208 μs</td><td>0.0194 μs</td><td>0.4845</td><td>-</td><td>-</td><td>2.98 KB</td>
</tr><tr><td>BulkCopyToSqlServer_cachedTarget_1</td><td>1,220.698 μs</td><td>8.6026 μs</td><td>8.0469 μs</td><td>1.9531</td><td>-</td><td>-</td><td>15.63 KB</td>
</tr><tr><td>BulkCopyToSqlServer_cachedTarget_9</td><td>1,194.385 μs</td><td>14.3305 μs</td><td>13.4047 μs</td><td>1.9531</td><td>-</td><td>-</td><td>17.09 KB</td>
</tr><tr><td>BulkCopyToSqlServer_cachedTarget_81</td><td>1,420.294 μs</td><td>13.2123 μs</td><td>12.3588 μs</td><td>3.9063</td><td>-</td><td>-</td><td>31.33 KB</td>
</tr><tr><td>BulkCopyToSqlServer_cachedTarget_729</td><td>3,124.885 μs</td><td>28.9503 μs</td><td>27.0801 μs</td><td>23.4375</td><td>-</td><td>-</td><td>157.48 KB</td>
</tr><tr><td>BulkCopyToSqlServer_uncachedTarget_0</td><td>99.438 μs</td><td>1.4363 μs</td><td>1.3435 μs</td><td>1.5869</td><td>-</td><td>-</td><td>10.45 KB</td>
</tr><tr><td>BulkCopyToSqlServer_uncachedTarget_1</td><td>1,271.974 μs</td><td>13.9532 μs</td><td>13.0518 μs</td><td>1.9531</td><td>-</td><td>-</td><td>23.09 KB</td>
</tr><tr><td>BulkCopyToSqlServer_uncachedTarget_9</td><td>1,301.108 μs</td><td>12.4396 μs</td><td>11.6360 μs</td><td>3.9063</td><td>-</td><td>-</td><td>24.57 KB</td>
</tr><tr><td>BulkCopyToSqlServer_uncachedTarget_81</td><td>1,505.734 μs</td><td>7.4808 μs</td><td>6.9976 μs</td><td>5.8594</td><td>-</td><td>-</td><td>38.81 KB</td>
</tr><tr><td>BulkCopyToSqlServer_uncachedTarget_729</td><td>3,218.563 μs</td><td>24.6221 μs</td><td>23.0315 μs</td><td>23.4375</td><td>-</td><td>-</td><td>164.99 KB</td>
</tr></tbody></table>

This branch (0d1798354aeaf20003f2a0643c69e9355e639fb2): 

<pre><code>
BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19041.508 (2004/?/20H1)
Intel Core i7-8700K CPU 3.70GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=5.0.100-rc.1.20452.10
  [Host] : .NET Core 3.1.8 (CoreCLR 4.700.20.41105, CoreFX 4.700.20.41903), X64 RyuJIT
</code></pre>
<pre><code>Job=InProcess  MaxRelativeError=0.002  OutlierMode=DontRemove  
Toolchain=InProcessNoEmitToolchain  IterationCount=15  LaunchCount=1  
MaxIterationCount=5000  WarmupCount=3  
</code></pre>

<table>
<thead><tr><th>                          Method</th><th>    Mean</th><th> Error</th><th>StdDev</th><th>Gen 0</th><th>Gen 1</th><th>Gen 2</th><th>Allocated</th>
</tr>
</thead><tbody><tr><td>BulkInsertTarget_LoadFromTable</td><td>116,304.1 ns</td><td>3,119.14 ns</td><td>2,917.64 ns</td><td>0.9766</td><td>-</td><td>-</td><td>7642 B</td>
</tr><tr><td>BulkCopyToSqlServer_cachedTarget_0</td><td>117.8 ns</td><td>2.39 ns</td><td>2.24 ns</td><td>0.0215</td><td>-</td><td>-</td><td>136 B</td>
</tr><tr><td>BulkCopyToSqlServer_cachedTarget_1</td><td>323,362.4 ns</td><td>17,250.52 ns</td><td>16,136.15 ns</td><td>2.4414</td><td>-</td><td>-</td><td>18074 B</td>
</tr><tr><td>BulkCopyToSqlServer_cachedTarget_9</td><td>1,246,746.4 ns</td><td>24,592.56 ns</td><td>23,003.90 ns</td><td>1.9531</td><td>-</td><td>-</td><td>18096 B</td>
</tr><tr><td>BulkCopyToSqlServer_cachedTarget_81</td><td>1,418,329.2 ns</td><td>15,300.09 ns</td><td>14,311.72 ns</td><td>3.9063</td><td>-</td><td>-</td><td>32676 B</td>
</tr><tr><td>BulkCopyToSqlServer_cachedTarget_729</td><td>3,114,785.2 ns</td><td>12,643.70 ns</td><td>11,826.93 ns</td><td>23.4375</td><td>-</td><td>-</td><td>161853 B</td>
</tr><tr><td>BulkCopyToSqlServer_uncachedTarget_0</td><td>90,180.4 ns</td><td>1,471.91 ns</td><td>1,376.82 ns</td><td>1.2207</td><td>-</td><td>-</td><td>7778 B</td>
</tr><tr><td>BulkCopyToSqlServer_uncachedTarget_1</td><td>401,740.7 ns</td><td>4,448.68 ns</td><td>4,161.29 ns</td><td>3.9063</td><td>-</td><td>-</td><td>25722 B</td>
</tr><tr><td>BulkCopyToSqlServer_uncachedTarget_9</td><td>1,294,510.9 ns</td><td>10,776.88 ns</td><td>10,080.70 ns</td><td>3.9063</td><td>-</td><td>-</td><td>25749 B</td>
</tr><tr><td>BulkCopyToSqlServer_uncachedTarget_81</td><td>1,521,271.2 ns</td><td>5,514.05 ns</td><td>5,157.85 ns</td><td>5.8594</td><td>-</td><td>-</td><td>40332 B</td>
</tr><tr><td>BulkCopyToSqlServer_uncachedTarget_729</td><td>3,227,109.2 ns</td><td>34,976.11 ns</td><td>32,716.67 ns</td><td>23.4375</td><td>-</td><td>-</td><td>169538 B</td>
</tr></tbody></table>

### CODE REVIEW

**Comment by EamonNerbonne** [comment]
Geen haast, wel interessant om ooit te bespreken; het is namelijk aantrekkelijk als we geen uitzonderingen hoeven maken voor inserts van "weinig" data.

**Comment by EamonNerbonne** [comment]
@fdij @chrarnoldus kunnen jullie hiernaar kijken?  De doelstelling is dat een losse insert nauwelijks nog voordeel bied boven een bulk insert, zodat we luier kunnen zijn (vooral ook in de testen) zonder dat dingen trager worden. Het verschilt toch nog wel een factor drie, dus ik denk dat dat niet helemaal zinloos is, gezien hoe vaak zijn toch enkele records inserten.

**Comment by chrarnoldus** [comment]
Hebben we een test voor meer dan 10 items?

**Comment by EamonNerbonne** [comment]
> Hebben we een test voor meer dan 10 items?

Zo niet moet die er natuurlijk komen.  Maar waar ik me vooral over zorgen maak is over de aanpak - is dit ueberhaupt een goed idee?  Bij welke bulk-insert-options moeten we dit doen?  De oude bulk-insert default had *geen* keep nulls, maar `insert into` gedraagd zich wel zo - moeten we/kunnen we dat zomaar aanpassen (ik zou liefst de scalar en bulk situaties nog even los van deze optimalisatie zo veel mogelijk op elkaar laten lijken, want anders is het zo'n gotcha).

**Comment by EamonNerbonne** [comment]
@chrarnoldus er waren wel wat tests die toevallig de non-small-batch pad raakten, maar niks expliciets, dus ik heb een extra test toegevoegd.

File: src/ProgressOnderwijsUtils/Pocos/BulkInsertTarget.cs [hunk 1]
```diff
@@ -40,10 +47,59 @@ public BulkInsertTarget With(BulkCopyFieldMappingMode mode)
...
+
+            return (completeHead, FullSequence());
+        }
+
         public void BulkInsert<[MeansImplicitUse(ImplicitUseKindFlags.Access, ImplicitUseTargetFlags.WithMembers)]
             T>(SqlConnection sqlConn, IEnumerable<T> pocos, CommandTimeout timeout = default, CancellationToken cancellationToken = default)
             where T : IReadImplicitly
         {
+            if (Options == defaultBulkCopyOptions) {
+                var (head, all) = PeekAtPrefix(pocos, 10);
```
Comment (tjibbechris) [inline src/ProgressOnderwijsUtils/Pocos/BulkInsertTarget.cs:95]: die 10 naar const of var?

File: src/ProgressOnderwijsUtils/Pocos/BulkInsertTarget.cs [hunk 2]
```diff
@@ -40,10 +47,59 @@ public BulkInsertTarget With(BulkCopyFieldMappingMode mode)
...
+            }
+
+            return (completeHead, FullSequence());
+        }
+
         public void BulkInsert<[MeansImplicitUse(ImplicitUseKindFlags.Access, ImplicitUseTargetFlags.WithMembers)]
             T>(SqlConnection sqlConn, IEnumerable<T> pocos, CommandTimeout timeout = default, CancellationToken cancellationToken = default)
             where T : IReadImplicitly
         {
+            if (Options == defaultBulkCopyOptions) {
```
Comment (tjibbechris) [inline src/ProgressOnderwijsUtils/Pocos/BulkInsertTarget.cs:94]: Dit als explicitie parameter?

**Comment by tjibbechris** [comment]
Ik zie geen bezwaar

File: src/ProgressOnderwijsUtils/Pocos/BulkInsertTarget.cs (hunk 1 above)
Comment (EamonNerbonne) [reply src/ProgressOnderwijsUtils/Pocos/BulkInsertTarget.cs:95]: Ik denk dat ik em ook ga tweaken naar iets lagers, maar idd, een const lijkt me goed.

File: src/ProgressOnderwijsUtils/Pocos/BulkInsertTarget.cs (hunk 2 above)
Comment (EamonNerbonne) [reply src/ProgressOnderwijsUtils/Pocos/BulkInsertTarget.cs:94]: Dit snap ik niet helemaal - wat wil je als expliciete parameter?  De defaults zijn echt de defaults; dat is een constante, en de Options, zijn whatever je hier hebt ingesteld (conceptueel een parameter, maar dit is een builder),

Comment (tjibbechris) [reply src/ProgressOnderwijsUtils/Pocos/BulkInsertTarget.cs:94]: Ik bedoel de hele bool als parameter. Ik vind de afhankelijkheid van een specifieke set Options niet heel voor de hand liggend

**Comment by chrarnoldus** [comment]
> ik zou liefst de scalar en bulk situaties nog even los van deze optimalisatie zo veel mogelijk op elkaar laten lijken, want anders is het zo'n gotcha

lijkt me prima.

File: test/ProgressOnderwijsUtils.Tests/Data/BulkInsertTest.cs [hunk 3]
```diff
@@ -74,35 +73,56 @@ sealed class SampleRow : ValueBase<SampleRow>, IWrittenImplicitly, IReadImplicit
...
-            var fromDb = SQL($"select * from #test").ReadPocos<SampleRow>(Connection);
-            AssertCollectionsEquivalent(SampleData, fromDb);
+            var target = BulkInsertTestSampleRow.CreateTable(Connection, SQL($"#test"));
+            BulkInsertTestSampleRow.SampleRows(4).BulkCopyToSqlServer(Connection, target);
+            var fromDb = SQL($"select * from #test").ReadPocos<BulkInsertTestSampleRow>(Connection);
+            AssertCollectionsEquivalent(BulkInsertTestSampleRow.SampleRows(4), fromDb);
+        }
+
+        [Fact]
+        public void BulkInsertAndReadRoundTrips_ManyRows()
```
Comment (chrarnoldus) [inline test/ProgressOnderwijsUtils.Tests/Data/BulkInsertTest.cs:100]: Hebben we nog een test nodig voor de andere situatie wat betreft `enumerable is IReadOnlyList<T> || enumerable is ICollection<T>`?

Comment (EamonNerbonne) [reply test/ProgressOnderwijsUtils.Tests/Data/BulkInsertTest.cs:100]: zijnde dat die gewoon de sequence onveranderd doorgeeft leek me dat minder boeiend.

Comment (EamonNerbonne) [reply test/ProgressOnderwijsUtils.Tests/Data/BulkInsertTest.cs:101]: maar goed, een minimale test is niet moeilijk, toegevoegd.

**Comment by EamonNerbonne** [comment]
Ok, ik heb de semantiek van de aanpak zoals jullie bekeken hebben behouden, maar toegevoegd zijn wat tests, en de implementatie is in een losse file gekomen zodat het overzichtelijker is, en ik heb KeepNull optioneel gemaakt, just in case (met test).

File: src/ProgressOnderwijsUtils/Pocos/BulkInsertTarget.cs (hunk 2 above)
Comment (EamonNerbonne) [reply src/ProgressOnderwijsUtils/Pocos/BulkInsertTarget.cs]: Ik heb het iets anders opgelost, maar ik conceptueel wil ik dit checken dichtbij de small-batch implementatie, want die heeft kennis van welke opties die ondersteund.  Als ik hier ergens ooit een bool van zou maken, dan moet de aanroeper weten wat allemaal wel werkt, en wat niet, en dat lijkt me tricky.  Zo is het een implementatie-detail die alleen relevant is als optimalisatie.

Comment (tjibbechris) [reply src/ProgressOnderwijsUtils/Pocos/BulkInsertTarget.cs]: Maar nu is het geschreven als een afhankelijkeheid van de defaultBulkCopyOptions, terwijl die afhankelijkheid er niet is

Comment (EamonNerbonne) [reply src/ProgressOnderwijsUtils/Pocos/BulkInsertTarget.cs]: Ik zou `defaultBulkCopyOptions` `optionsThatMirrorSingleRowInsertIntoBehavior` oid; maar iig die afhankelijkheid is niet toevallig.  De defaults zijn gekozen om het verschil tussen `insert into` en bulk-copy zo klein mogelijk te houden, en deze optimalisatie werkt alleen als de code aan kan nemen dat insert into en bulk copy hetzelfde doen (bijv. in termen van fire triggers enzo).

Comment (EamonNerbonne) [reply src/ProgressOnderwijsUtils/Pocos/BulkInsertTarget.cs]: renamed.

File: src/ProgressOnderwijsUtils/Pocos/SmallBatchInsertImplementation.cs [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+namespace ProgressOnderwijsUtils
+{
+    public static class SmallBatchInsertImplementation
+    {
+        internal const int ThresholdForUsingSqlBulkCopy = 6;
+
+        public static IEnumerable<T>? TrySmallBatchInsertOptimization<T>(SqlConnection sqlConn, BulkInsertTarget bulkInsertTarget, IEnumerable<T> pocos, CommandTimeout timeout)
+            where T : IReadImplicitly
+        {
+            if ((bulkInsertTarget.Options | SqlBulkCopyOptions.KeepNulls) != (BulkInsertTarget.DefaultOptionsCorrespondingToInsertIntoBehavior | SqlBulkCopyOptions.KeepNulls)) {
```
Comment (tjibbechris) [inline src/ProgressOnderwijsUtils/Pocos/SmallBatchInsertImplementation.cs:19]: Die ``KeepNulls`` zitten toch al in DefaultOptionsCorrespondingToInsertIntoBehavior?

Comment (EamonNerbonne) [reply src/ProgressOnderwijsUtils/Pocos/SmallBatchInsertImplementation.cs:19]: ja, maar nu zijn ze niet meer nodig - mag aan, mag uit.

File: test/ProgressOnderwijsUtils.Tests/Data/BulkInsertTest.cs [hunk 5]
```diff
@@ -161,24 +186,73 @@ static void AssertCollectionsEquivalent<T>(T[] sampleData, T[] fromDb)
...
+                        , stringNull nvarchar(max) null
+                        , stringNonNull nvarchar(max) not null
+                    )
+                "
+            ).ExecuteNonQuery(Connection);
+            var target = BulkInsertTarget.LoadFromTable(Connection, "#tmp");
+            new[] { new SampleRow2 { intNonNull = 1, intNull = null, stringNull = "test", stringNonNull = "test" }, }
+                .BulkCopyToSqlServer(Connection, target);
+            new[] { new SampleRow2 { intNonNull = 2, intNull = null, stringNull = "test", stringNonNull = "test" }, }
+                .BulkCopyToSqlServer(Connection, target.With(target.Options ^ SqlBulkCopyOptions.KeepNulls));
```
Comment (EamonNerbonne) [inline test/ProgressOnderwijsUtils.Tests/Data/BulkInsertTest.cs:206]: @tjibbechris deze test doet een keer met en een keer zonder keep nulls, om te checken dat beide werkt.

File: src/ProgressOnderwijsUtils/Pocos/SmallBatchInsertImplementation.cs [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+            if (maybeMapping.IsError) {
+                throw new InvalidOperationException($"Failed to map source {typeof(T).ToCSharpFriendlyTypeName()} to the table {target.TableName}. Errors:\r\n{maybeMapping.AssertError()}");
+            }
+            var mapping = maybeMapping.AssertOk();
+            foreach (var row in rows) {
+                var destinationColumns = mapping.Select(o => ParameterizedSql.CreateDynamic(o.Dst.Name));
+                var sourceValues = mapping.Select(
+                    o => {
+                        var fieldVal = PocoProperties<T>.Instance[o.Src.Index].Getter.AssertNotNull()(row);
+                        return fieldVal == null && !target.Options.HasFlag(SqlBulkCopyOptions.KeepNulls) ? SQL($"default") : ParameterizedSql.Param(fieldVal);
```
Comment (EamonNerbonne) [inline src/ProgressOnderwijsUtils/Pocos/SmallBatchInsertImplementation.cs:48]: En hier, indien not-KeepNulls, wordt expliciet de default ge-insert.

File: src/ProgressOnderwijsUtils/Pocos/SmallBatchInsertImplementation.cs (hunk 4 above)
Comment (tjibbechris) [reply src/ProgressOnderwijsUtils/Pocos/SmallBatchInsertImplementation.cs:19]: Ik snap wel dat het werkt, maar ik vind het niet heel leesbaar wat deze regel doet.

Comment (EamonNerbonne) [reply src/ProgressOnderwijsUtils/Pocos/SmallBatchInsertImplementation.cs:19]: Heb je een beter voorstel?

Comment (EamonNerbonne) [reply src/ProgressOnderwijsUtils/Pocos/SmallBatchInsertImplementation.cs:19]:
```suggestion
            if (((BulkInsertTarget.DefaultOptionsCorrespondingToInsertIntoBehavior ^ bulkInsertTarget.Options) & ~SqlBulkCopyOptions.KeepNulls) != default) {
```
is dit beter?

Comment (tjibbechris) [reply src/ProgressOnderwijsUtils/Pocos/SmallBatchInsertImplementation.cs:19]: misschien is een commentaarregel die duidelijkmaakt waarom deze escape er is wel voldoende

Comment (EamonNerbonne) [reply src/ProgressOnderwijsUtils/Pocos/SmallBatchInsertImplementation.cs:19]:
```suggestion
            if ((bulkInsertTarget.Options | SqlBulkCopyOptions.KeepNulls) != (BulkInsertTarget.DefaultOptionsCorrespondingToInsertIntoBehavior | SqlBulkCopyOptions.KeepNulls)) {
                // alleen heel specifieke options gedragen zich hetzelfde als "insert into", dus als iemand afwijkende options heeft gekozen: abort.
```

**Review by tjibbechris — APPROVED** [review]

**Pull request merged**